src/utils/ast/astUtils.ts

```diff
@@ -0,0 +1,106 @@
+    },
+    Identifier: (id: es.Identifier) => ids.add(id),
+    ObjectPattern: ({ properties }: es.ObjectPattern, _state, c) =>
+      properties.forEach(prop => c(prop, null)),
+    RestElement: ({ argument }: es.RestElement, _state, c) => c(argument, null)
+  })
+  return ids
+}
+
+export function declarationToExpression(node: es.ClassDeclaration): es.ClassExpression
+export function declarationToExpression(node: es.FunctionDeclaration): es.FunctionExpression
+export function declarationToExpression({
+  type,
+  ...node
+}: es.FunctionDeclaration | es.ClassDeclaration) {
+  return {
+    ...node,
+    type: type === 'FunctionDeclaration' ? 'FunctionExpression' : 'ClassExpression'
+  } as es.FunctionExpression | es.ClassExpression
+}
+
+type ExportDefaultProcessors<T> = {
+  FunctionDeclaration: (node: es.FunctionDeclarationWithId) => T
+  ClassDeclaration: (node: es.ClassDeclarationWithId) => T
+  Expression: (node: es.Expression) => T
+}
+
+export function processExportDefaultDeclaration<T>(
+  node: es.ExportDefaultDeclaration,
+  processors: ExportDefaultProcessors<T>
+) {
+  if (isDeclaration(node.declaration)) {
+    const declaration = node.declaration
+    assert(
+      declaration.type !== 'VariableDeclaration',
+      'ExportDefaultDeclarations cannot have VariableDeclarations'
+    )
+
+    if (declaration.type === 'FunctionDeclaration') {
+      if (declaration.id) {
+        return processors.FunctionDeclaration(declaration as es.FunctionDeclarationWithId)
+      }
+
+      return processors.Expression({
+        ...declaration,
+        type: 'FunctionExpression'
+      })
+    }
+
+    if (declaration.id) {
+      return processors.ClassDeclaration(declaration as es.ClassDeclarationWithId)
+    }
+
+    return processors.Expression({
+      ...declaration,
+      type: 'ClassExpression'
+    })
+  }
+
+  return processors.Expression(node.declaration)
+}
+
+type ExportNamedProcessors<T> = {
+  withVarDecl: (node: es.VariableDeclaration) => T
+  withClass: (node: es.ClassDeclarationWithId) => T
+  withFunction: (node: es.FunctionDeclarationWithId) => T
+  localExports: (node: es.ExportNamedLocalDeclaration) => T
+  withSource: (node: es.ExportNamedDeclarationWithSource) => T
+}
+
+export function processExportNamedDeclaration<T>(
+  node: es.ExportNamedDeclaration,
+  processors: ExportNamedProcessors<T>
+) {
+  if (node.declaration) {
+    switch (node.declaration.type) {
+      case 'VariableDeclaration':
+        return processors.withVarDecl(node.declaration)
+      case 'FunctionDeclaration':
+        return processors.withFunction(node.declaration as es.FunctionDeclarationWithId)
+      case 'ClassDeclaration':
+        return processors.withClass(node.declaration as es.ClassDeclarationWithId)
+    }
+  } else if (node.source) {
+    return processors.withSource(node as es.ExportNamedDeclarationWithSource)
+  } else {
+    return processors.localExports(node as es.ExportNamedLocalDeclaration)
+  }
+}
```

Comment:
How do you know that all of these type assertions are safe? One huge issue I faced while working on my FYP was the use of (incorrect) type assertions and non-null assertion operators everywhere that bypassed type safety. It meant that while the application could compile, I had a lot of trouble fixing the runtime errors that occurred due to incorrect types because I cannot possibly know where errors occur until I run the program.

From a maintainability point of view, this is a blocker for me.

Reply:
I actually encountered the same problems you did, which is why I created these functions. The different variations of each type of export node weren't being properly handled, so to keep to DRY and make the handing more robust I added these functions.

For the two you've highlighted here, if you follow the code flow, they account for every single variant (accounted for by Typescript) of both the `ExportDefaultDeclaration` and `ExportNamedDeclaration`.

Reply:
For the 2 type assertions above, could you make use of [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) which contain only the relevant code that lets us know that the type assertion is safe. There are 2 reasons for this:
1. We avoid using type assertions with the long term goal of eventually banning them via a lint rule (unless suppressed along with an explanation).
1. By encapsulating all of the logic that helps us check if a type conversion is valid within a function, it is easier to verify, and we can use the type predicate in other places of the codebase.

The rest of the type assertions within the code snippet (within `processExportNamedDeclaration`) look unsafe to me though.